docs/conditions.rst

```diff
@@ -249,3 +251,172 @@ do not need actual record by are only looking to traverse::
+
+Creates a CompoundCondition object from an array::
+
+	// below will create 2 conditions and nest them in a compound conditions with AND junction
+	$compoundCondition1 = new CompoundCondition([
+		['name', 'like', 'ABC%'],
+		['country', 'US']
+	]);
+	
+.. php:method:: negate();
+
+Negate method has behind the full map of conditions so any condition object can be negated, e.g negating '>=' results in '<', etc.
+For compound conditionss this method is using De Morgan's laws, e.g::
+
+	// using $compoundCondition1 defined above
+	// results in "(Name not like 'ABC%') or (Country does not equal 'US')"
+	$compoundCondition1->negate();
+
+.. php:method:: mergeAnd(AbstractCondition $conditionA, AbstractCondition $conditionB, $_ = null);
+
+Merge number of conditions using AND as junction. Returns the resulting CompoundCondition object.
+
+.. php:method:: mergeOr(AbstractCondition $conditionA, AbstractCondition $conditionB, $_ = null);
+
+Merge number of conditions using OR as junction. Returns the resulting CompoundCondition object.
+
+.. php:method:: simplify();
+
+Peels off single nested conditions. Useful for converting (((field = value))) to field = value.
+
+.. php:method:: clear();
+
+Clears the condition from nested conditions.
+
+.. php:method:: isOr();
+
+Checks if scope components are joined by OR
+
+.. php:method:: isAnd();
+
+Checks if scope components are joined by AND
+
+Conditions on Referenced Models
+-------------------------------
+
+Agile Data allows for adding conditions on related models for retrieval of type 'model has references where'.
+
+Setting conditions on references can be done utilizing the Model::refLink method but there is a shorthand format 
+directly integrated with addCondition method using "/" to chain the reference names::
+
+	$contact->addCondition('company/country', 'US');
+	
+This will limit the $contact model to those whose company is in US.
+'company' is the name of the reference in $contact model and 'country' is a field in the referenced model.
+
+If a condition must be set directly on the existence or number of referenced records the special symbol "#" can be
+utilized to indicate the condition is on the number of records::
+
+	$contact->addCondition('company/tickets/#', '>', 3);
+	
+This will limit the $contact model to those whose company have more than 3 tickets.
+'company' and 'tickets' are the name of the chained references ('company' is a reference in the $contact model and
+'tickets' is a reference in Company model)
+
+For applying conditions on existence of records the '?' (has any) and '!' (doesn't have any) special symbols can be used.
+Although it is similar in functionality to checking ('company/tickets/#', '>', 0) or ('company/tickets/#', '=', 0)
+'?' and '!' special symbols use optimized query and are much faster::
+
+	// Contact whose company has any tickets
+	$contact->addCondition('company/tickets/?');
```

Comment:
I think this should be not supported

with `#`, we can get query count and apply regular condition to it

and internally optimize checks of `#` > 0 to exist

`#` seems like a magical column name to me - probably ok
`?`, ... seems like a n things mixed together, actually, and, if we support unary condition (do we?, ie. without operator and value), then `#` has the same meaning

Reply:
We have 3 special symbols:
'#' -> has # referenced records
'?' -> has any referenced records
'!' -> does not have any referenced records

The '?' and '!' use Persistence `exists` action for performance purposes so they make sense.

Reply:
> and internally optimize checks of # > 0 to exist

that is why I think we do not needed, as we should optimize `#` anyway and offer exactly the same with unified design

we can even futher optimize like `(SELECT COUNT(*) FROM (SELECT 1 FROM t WHERE ... LIMIT 2)) >= 2` when the condition is `#` > 1 :)

Reply:
Let's start simple and then optimize if possible. I consider using the 2 additional special symbols a good solution for the extreme cases including optimization.

Reply:
> Let's start simple and then optimize if possible.

yes - I propose `#` only and optimize at least `= 1`, `> 0` and `>= 1` to "like ? before", and `= 0` to "like ! before"

Reply:
Why do you want to remove the other special symbols?
"Everything should be made as simple as possible, but no simpler" :)

Reply:
get rid of `?` and `!` special meaning characters

`#` is special enought

this is important for simple and understandable design and also for more freedom in the future

Reply:
I don't see anything bad if we use 3 special symbols, but of course it would be good if under-the-hood toolkit could optimize `#>0` to `?` and use `exists` (and other cases mentioned above).
Also need to remember that in theory we could have some persistence which doesn't support `exists` or `limit`. For example, some `Persistence_RestApi` or something like that. In that case also having multiple special symbols could (maybe) help somehow.
Anyway as I said - I don't see anything bad about having them.

Reply:
We was discussing this with @mvorisek for about hour and came to conclusion that:
- should use only #
- should not use ? and !. Instead internally need to optimize and use `exists` in case of `#>0` and `#=0`. In future also could optimize `#>n` to something like `(SELECT COUNT(*) FROM (SELECT 1 FROM t WHERE ... LIMIT n)) >= n`